picard/webservice.py

```diff
@@ -134,6 +135,87 @@ def _read_xml(stream):
     return document
 
 
+class WsRequest(object):
+
+    def __init__(self, method, host, port, path, data, handler, xml,
+                 mblogin=False, cacheloadcontrol=None, refresh=None,
+                 access_token=None, queryargs=None, retries=0,
+                 priority=False, important=False):
+        self.method = method
+        self.host = host
+        self.port = port
+        self.path = path
+        self.data = data
+        self.handler = handler
+        self.xml = xml
+        self.mblogin = mblogin
+        self.cacheloadcontrol = cacheloadcontrol
+        self.refresh = refresh
+        self.access_token = access_token
+        self.queryargs = queryargs
+        self.priority = priority
+        self.important = important
+        self.retries = retries
+
+        self._qrequest = None
+
+    def get_host_key(self):
+        return (self.host, self.port)
+
+    @property
+    def qrequest(self):
+        if self._qrequest is None:
+            url = build_qurl(self.host, self.port, path=self.path,
+                             mblogin=self.mblogin,
+                             queryargs=self.queryargs)
+            self._qrequest = QtNetwork.QNetworkRequest(url)
+        return self._qrequest
+
+    def send(self, request_methods):
+        if self.mblogin and self.access_token:
+            self.qrequest.setRawHeader("Authorization", "Bearer %s" % self.access_token)
+        if self.mblogin or (self.method == "GET" and self.refresh):
+            self.qrequest.setPriority(QtNetwork.QNetworkRequest.HighPriority)
+            self.qrequest.setAttribute(QtNetwork.QNetworkRequest.CacheLoadControlAttribute,
+                                 QtNetwork.QNetworkRequest.AlwaysNetwork)
+        elif self.method in ("PUT", "DELETE"):
+            self.qrequest.setPriority(QtNetwork.QNetworkRequest.HighPriority)
+        elif self.cacheloadcontrol is not None:
+            self.qrequest.setAttribute(QtNetwork.QNetworkRequest.CacheLoadControlAttribute,
+                                 self.cacheloadcontrol)
+        self.qrequest.setRawHeader("User-Agent", USER_AGENT_STRING)
+        if self.xml:
+            self.qrequest.setRawHeader("Accept", "application/xml")
+        if self.data is not None:
+            if (self.method == "POST"
+                and self.host == config.setting["server_host"]
+                and self.xml):
+                self.qrequest.setHeader(QtNetwork.QNetworkRequest.ContentTypeHeader, "application/xml; charset=utf-8")
+            else:
+                self.qrequest.setHeader(QtNetwork.QNetworkRequest.ContentTypeHeader, "application/x-www-form-urlencoded")
+        send = request_methods[self.method]
```

Comment:
Nitpick: this could use a different name than `send` because it's created in a method called `send` already.